libs/parallelism/futures/include/hpx/futures/future.hpp

```diff
@@ -782,8 +781,91 @@ namespace hpx { namespace lcos {
+        template <template <typename...> class Tuple,
+            template <typename...> class Variant>
+        using value_types = std::conditional_t<std::is_void<result_type>::value,
+            Variant<Tuple<>>,
+            Variant<Tuple<std::add_rvalue_reference_t<result_type>>>>;
+
+        template <template <typename...> class Variant>
+        using error_types = Variant<std::exception_ptr>;
+
+        static constexpr bool sends_done = false;
+
+        template <typename Receiver>
+        struct operation_state
+        {
+            std::decay_t<Receiver> receiver;
+            std::decay_t<future> fut;
+
+            operation_state(operation_state&&) = delete;
+            operation_state(operation_state const&) = delete;
+
+            void start_helper(std::true_type) noexcept
+            {
+                try
+                {
+                    fut.then([receiver = std::move(receiver)](
+                                 future&& fut) mutable {
+                        if (fut.has_value())
+                        {
+                            hpx::execution::experimental::set_value(
+                                std::move(receiver));
+                        }
+                        else if (fut.has_exception())
+                        {
+                            hpx::execution::experimental::set_error(
+                                std::move(receiver), fut.get_exception_ptr());
+                        }
+                    });
+                }
+                catch (...)
+                {
+                    hpx::execution::experimental::set_error(
+                        std::move(receiver), std::current_exception());
+                }
+            }
+
+            void start_helper(std::false_type) noexcept
+            {
+                try
+                {
+                    fut.then([receiver = std::move(receiver)](
+                                 future&& fut) mutable {
+                        if (fut.has_value())
+                        {
+                            hpx::execution::experimental::set_value(
+                                std::move(receiver), fut.get());
+                        }
+                        else if (fut.has_exception())
+                        {
+                            hpx::execution::experimental::set_error(
+                                std::move(receiver), fut.get_exception_ptr());
+                        }
+                    });
+                }
+                catch (...)
+                {
+                    hpx::execution::experimental::set_error(
+                        std::move(receiver), std::current_exception());
+                }
+            }
+
+            void start() noexcept
+            {
+                start_helper(std::is_void<result_type>{});
+            }
+        };
+
+        template <typename Receiver>
+        operation_state<Receiver> connect(Receiver&& receiver) &&
```

Comment:
Does `connect` have to be a member function? Making it a global function might allow to disentangle the `operations_state` defined here from the future itself (we might have to make it a friend, though).

Reply:
No, it can be a tag_invoke overload. The only part that has to stay are the typedefs (`value_types` and `error_types`), but the rest can be outside the class.

Reply:
I moved the `operation_state` type outside the future classes, and there's only one (templated) type for `future` and `shared_future` now (instead of duplicated inside `future` and `shared_future`). I left it in this file though, because the `tag_invoke` overloads require the definition of the tag `connect_t` which would be problematic in terms of dependencies between modules.